File: pyton/15puzzle.py
"""
Note p: for permutation ex: [1, 2, 3, 4, 5, 6, 0, 7, 8]
"""

from time import time_ns
from math import sqrt
from random import shuffle
from queue import PriorityQueue


def is_even(p: list) -> bool:
    inversions = 0
    for i in range(len(p) - 1):
        for j in range(i, len(p)):
            if p[i] > p[j] and p[j] != 0:
                inversions += 1

    return inversions % 2 == 0


def manhatenDist(p: int) -> int:
    SIZE = int(sqrt(len(p)))
    dist = 0

    for i, el in enumerate(p):
        if (el != 0):
            row, col = (el-1) // SIZE, (el-1) % SIZE
            goal_row, goal_col = i // SIZE, i % SIZE

            dist += abs(row - goal_row) + abs(col - goal_col)

    return dist


def applyMoves(p: list, blank: int, path: list) -> list:
    res = []
    l = int(sqrt(len(p)))
    MOVES = [1, -1, l, -l]

    if blank % l == 0:
        MOVES = [1, l, -l]
    elif (blank + 1) % l == 0:
        MOVES = [-1, l, -l]

    for move in MOVES:
        if move + blank < len(p) and move + blank >= 0:
            copy = p.copy()
            changed = copy[move + blank]
            copy[move + blank], copy[blank] = copy[blank], copy[move + blank]
            res.append((
                manhatenDist(copy) + len(path) + 1,
                path + [move + blank],
                copy,
                move + blank
            ))

    return res


def getKey(p: list):
    return "".join(list(map(lambda x: str(x), p)))


def solution(p: list) -> list:
    if not is_even(p):
        return False

    visited = {}
    q = PriorityQueue()
    q.put((0, [p.index(0)], p, p.index(0)))
    goal = getKey(list(range(1, len(p))) + [0])

    while not q.empty():
        _, path, curr, blank = q.get()

        if getKey(curr) == goal:
            return path

        if getKey(curr) not in visited:
            visited[getKey(curr)] = True
            for nextState in applyMoves(curr, blank, path):
                q.put(nextState)


if __name__ == '__main__':
    p = list(range(1, 16))
    shuffle(p)
    p = p + [0]
    p = [ 6, 4, 0, 5, 8, 3, 1, 7, 2 ]
    print(is_even(p))
    print(p)
    start = time_ns()
    print(solution(p))
    print(str((time_ns() - start) // 1000000) + " ms")

File: js/15puzzle.js
const PriorityQueue = require("./tinyqueue.js").default

const getKey = p => p.join("")

const getGoal = p => getKey(
    new Array(p.length).fill(0)
        .map((_, i) => (i < p.length - 1) ? i + 1 : 0)
)

const isEven = p => {
    let inversions = 0

    for (let i = 0; i < p.length - 1; i++) {
        for (let j = i; j < p.length; j++) {
            if (p[i] > p[j] && p[j] !== 0) inversions += 1
        }
    }

    return inversions % 2 === 0
}

const manhatenDist = p => {
    const SIZE = parseInt(Math.sqrt(p.length), 10)

    return p
        .reduce(
            (dist, el, i) =>
                el !== 0 ?
                    // Accumulator Distance
                    dist +
                    // Col-Difference
                    Math.abs(((el - 1) % SIZE) - (i % SIZE)) +
                    // Row-Difference
                    Math.abs(parseInt((el - 1) / SIZE, 10) - parseInt(i / SIZE, 10))
                    :
                    dist, 0
        )
}

const applyMoves = (p, path, blank) => {
    const l = parseInt(Math.sqrt(p.length), 10)
    const MOVES =
        blank % l === 0 ? [1, l, -l] :
            (blank + 1) % l === 0 ?
                [-1, l, -l] : [1, -1, l, -l]

    return (
        MOVES
            // Remove illegal options diriction = 0 || diriction out of the board
            .filter(
                move => (move + blank < p.length && move + blank >= 0)
            )
            // Map to the new states
            .map(move => {
                const newBoard = p.slice(0)
                if (newBoard) [newBoard[move + blank], newBoard[blank]] = [newBoard[blank], newBoard[move + blank]]

                return [
                    // manhaten distance to detirmine the priority
                    manhatenDist(newBoard),
                    // The path of the blank so far
                    [...path, move + blank],
                    // new board
                    newBoard
                ]
            })
    )
}


const solution = p => {
    if (isEven(p)) {
        const visited = {}
        const q = new PriorityQueue([[0, [p.indexOf(0)], p]], (a, b) => a[0] - b[0])
        const goal = getGoal(p)

        while (q.length) {
            const [, path, curr] = q.pop()

            if (getKey(curr) === goal) return path

            if (!visited[getKey(curr)]) {
                visited[getKey(curr)] = true
                applyMoves(curr, path, path[path.length - 1]).forEach(nextState =>
                    q.push(nextState)
                )
            }
        }
    }
}

const p = shuffle(new Array(15).fill(0)
    .map((_, i) => i + 1))

console.log(JSON.stringify(solution([...p, 0])))

function shuffle(a) {
    var j, x, i;
    for (i = a.length - 1; i > 0; i--) {
        j = Math.floor(Math.random() * (i + 1));
        x = a[i];
        a[i] = a[j];
        a[j] = x;
    }
    return a;
}